src/Optimizer/DeriveRequiredChildProp.cpp

```diff
@@ -0,0 +1,316 @@
+#include <Optimizer/DeriveRequiredChildProp.h>
+#include <Optimizer/Group.h>
+
+
+namespace DB
+{
+
+void setSortProp(QueryPlanStepPtr step, SortProp & required_sort_prop, size_t child_index = 0)
+{
+    if (child_index >= step->getInputStreams().size())
+        return;
+
+    /// There are some cases where sort desc is misaligned with the header, and in this case it is not required to keep the sort prop.
+    /// E.g select * from aaa_all where name in (select name from bbb_all where name like '%d%') order by id limit 13 SETTINGS allow_experimental_query_coordination = 1, allow_experimental_analyzer = 1;
+    /// CreatingSetsStep header is id_0, name_1 but its sort desc is id
+    const auto & header = step->getInputStreams()[child_index].header;
+    for (const auto & sort_column : step->getInputStreams()[child_index].sort_description)
+        if (!header.has(sort_column.column_name))
+            return;
+
+    if (!step->getInputStreams()[child_index].sort_description.empty())
+    {
+        required_sort_prop.sort_description = step->getInputStreams()[child_index].sort_description;
+        required_sort_prop.sort_scope = step->getInputStreams()[child_index].sort_scope;
+    }
+}
+
+AlternativeChildrenProp DeriveRequiredChildProp::visit(QueryPlanStepPtr step)
+{
+    if (group_node->hasRequiredChildrenProp())
+        return group_node->getRequiredChildrenProp();
+
+    return Base::visit(step);
+}
+
+AlternativeChildrenProp DeriveRequiredChildProp::visitDefault(IQueryPlanStep & step)
+{
+    if (step.stepType() == StepType::Scan)
+        throw Exception(
+            ErrorCodes::NOT_IMPLEMENTED,
+            "Engine {} not implemented with CBO optimizer, please disable it by setting allow_experimental_query_coordination=0",
```

Comment:
afaik ReadFromRemoteParallelReplicas, ReadFromPreparedSource are fairly used for queries and i couldn't use them with the query coordination. query: 
SELECT
    count(*),
    hostName() AS h
FROM test_db.test_dist
GROUP BY h

with allow_experimental_parallel_reading_from_replicas set to 0(for ReadFromPreparedSource) and 2(for ReadFromRemoteParallelReplicas)

Reply:
Now we only support MergeTree tables. The reason is that we transfer query plan by SQL text right now, and  the other nodes receive the SQL text, then will re-compile it and fetch the proper part of the plan.

Let's assume that we have 2 nodes, table in node1 is blank  and not in node2, then they will get different query plan which may lead query failure.

We left ser-des query plan in our TODO list, and after it we can support nearly all storages.

Reply:
Maybe there is another solution, just disable `ReadFromPreparedSource` temporarily.